src/components/legend/index.js

```diff
@@ -663,14 +686,53 @@ legend.repositionLegend = function(td, traces){
     lx = Math.round(lx);
     ly = Math.round(ly);
 
-    fullLayout._infolayer.selectAll('svg.legend')
-        .call(Plotly.Drawing.setRect, lx, ly, legendwidth, legendheight);
-    fullLayout._infolayer.selectAll('svg.legend .bg')
-        .call(Plotly.Drawing.setRect, borderwidth/2, borderwidth/2,
-            legendwidth-borderwidth, legendheight-borderwidth);
+
+    var legendsvg = fullLayout._infolayer.selectAll('svg.legend'),
+        scrollBox = fullLayout._infolayer.selectAll('svg.legend .scrollbox'),
+        scrollBar = fullLayout._infolayer.selectAll('svg.legend .scrollbar'),
+        bg = fullLayout._infolayer.selectAll('svg.legend .bg');
+
+    var plotHeight = fullLayout.height - fullLayout.margin.t - fullLayout.margin.b,
+        scrollheight = Math.min(plotHeight - ly, legendheight),
+        scrollPosition = scrollBox.attr('viewBox')? scrollBox.attr('viewBox').split(' ')[1] : 0;
+
+    legendsvg.node().addEventListener('wheel', scrollHandler);
+    legendsvg.call(Drawing.setRect, lx, ly, legendwidth, scrollheight);
+
+    bg.style({ width: legendwidth, height: scrollheight });
+
+    scrollBox.attr('viewBox', '0 ' + scrollPosition +' ' + legendwidth + ' ' + scrollheight);
+
+    if(td.firstRender) scrollBar.call(Drawing.setRect, legendwidth - 6, 10, 4, 20);
+
+    function scrollHandler(e){
+        e.preventDefault();
+
+        // Scale movement to simulate native scroll performance
+        var scrollDiff = e.deltaY / 25,
+            viewBox = scrollBox.attr('viewBox').split(' ');
+
+        var scrollBoxY = constrain(0, Math.max(legendheight - scrollheight, 0), +viewBox[1] + scrollDiff),
+            scrollBarY = scrollBoxY / legendheight * (scrollheight) + 10;
+
+        viewBox[1] = scrollBoxY;
+
+        scrollBox.attr('viewBox', viewBox.join(' '));
+        scrollBar.call(Drawing.setRect, legendwidth - 6, scrollBarY, 4, 20);
+    }
+
+    function constrain(min, max, c){
```

Comment:
[`Lib.constrain`](https://github.com/plotly/plotly.js/blob/master/src/lib/index.js#L125) will do just that.